File: app/src/main/java/com/pixeldraw/dbrt/pixeldraw/PixelPicView.java
package com.pixeldraw.dbrt.pixeldraw;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.ColorSpace;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.LayerDrawable;
import android.os.Build;
import android.support.annotation.Nullable;
import android.support.annotation.RequiresApi;
import android.support.design.widget.CoordinatorLayout;
import android.util.AttributeSet;
import android.util.Log;
import android.view.*;
import android.widget.FrameLayout;

import java.util.ArrayList;
import java.util.Scanner;

import static com.pixeldraw.dbrt.pixeldraw.AppGlobalData.MA_INSTANCE;
import static com.pixeldraw.dbrt.pixeldraw.AppGlobalData.Plates;

import static android.content.ContentValues.TAG;
@RequiresApi(api = Build.VERSION_CODES.O)
public class PixelPicView extends View {
    private Canvas canvas=new Canvas();
    private int widthPixels=16,heightPixels=16,webWidth=18;
    private int[][] Plate=new int[][]{{0}};
    private boolean isWeb=false;
    private Drawable background;
    private Bitmap bitmap;
    private FrameLayout.LayoutParams l;
    public int mod=0;
    public abstract static class OnPixelClickListener{
        public OnPixelClickListener(){ }
        public void onClick(View view,int x,int y) { }
    }
    public abstract static class OnPixelTouchListener{
        public OnPixelTouchListener(){ }
        public void onTouch(View view,MotionEvent motionEvent,int x,int y) { }
    }
    protected void updateSize(){
        Plate=new int[widthPixels][heightPixels];
        for (int i = 0; i < Plate.length; i++)
            for (int i1 = 0; i1 < Plate[i].length; i1++)
                Plate[i][i1]= Color.TRANSPARENT;
        l= (FrameLayout.LayoutParams) getLayoutParams();
        l.width=MA_INSTANCE.dip2px(300);
        l.height= ((int) (MA_INSTANCE.dip2px(300) * ((float) heightPixels / (float) widthPixels)));
        l.gravity=Gravity.CENTER;
        setX(MA_INSTANCE.displayMetrics.widthPixels/2-getMeasuredWidth()/2);
        setY(MA_INSTANCE.displayMetrics.heightPixels/2-getMeasuredHeight()/2);
        setScaleX(1);
        setScaleY(1);
        setLayoutParams(l);
        updateCanvas();
    }

    protected void updateCanvas() {
        invalidate();
    }
    public PixelPicView(Context ctx){
        super(ctx);
    }
    public PixelPicView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }
    public PixelPicView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }
    @Override
    public void invalidate() {
        super.invalidate();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        Paint mpaint2 = new Paint();
        mpaint2.setAntiAlias(false);
        mpaint2.setColor(Color.WHITE);
        for (int i = 0; i < Plate.length; i++)
            for (int i1 = 0; i1 < Plate[i].length; i1++) {
                Paint mpaint = new Paint();
                mpaint.setAntiAlias(false);
                mpaint.setColor(Plate[i][i1]);
                canvas.drawRect(this.getMeasuredWidth() / Plate.length * i, this.getMeasuredHeight() / Plate[0].length * i1, this.getMeasuredWidth() / Plate.length * (i + 1), this.getMeasuredHeight() / Plate[0].length * (i1 + 1), mpaint);
            }
        if(isWeb) {
            for (int i = 1; i < Plate.length+1; i++)
                canvas.drawRect(this.getMeasuredWidth() / Plate.length * i, 0f, this.getMeasuredWidth() / Plate.length * i + (getMeasuredWidth() / getWidthPixels() * 0.05f), this.getMeasuredHeight(), mpaint2);
            for (int i = 1; i < Plate[0].length+1; i++)
                canvas.drawRect(0f,this.getMeasuredHeight() / Plate[0].length * i - (getMeasuredHeight() / getHeightPixels() * 0.05f),getMeasuredWidth(),this.getMeasuredHeight() / Plate[0].length * i, mpaint2);

        }
        Rect rect=canvas.getClipBounds();
        Paint paint=new Paint();
        paint.setColor(Color.WHITE);
        paint.setStyle(Paint.Style.STROKE);
        paint.setStrokeWidth(4f);
        canvas.drawRect(rect,paint);
        super.onDraw(canvas);
    }
    public void enableWeb(boolean b){
        if(isWeb!=b) {
            isWeb = b;
            Log.d("c",""+b);
            updateCanvas();
        }
    }
    public void setWebWidth(int webWidth){
        this.webWidth=webWidth;
        updateCanvas();
    }
    public void setWidthPixels(int count){
        widthPixels=count;
        updateSize();
    }
    public void setHeightPixels(int count){
        heightPixels=count;
        updateSize();
    }
    public int getWidthPixels(){
        return widthPixels;
    }
    public int getHeightPixels(){
        return heightPixels;
    }

    public void set(int x, int y, int value){
        if(x>=0&&x<widthPixels&&y>=0&&y<heightPixels)
            if(Plate[x][y]!=value)
                Plate[x][y]=value;
        invalidate();
    }
    public void set(float x, float y, int value){
        if(x>=0&&x<widthPixels&&y>=0&&y<heightPixels)
            if(Plate[(int)x][(int)y]!=value)
                Plate[(int)x][(int)y]=value;
        invalidate();
    }
    public int get(int x,int y){
        if(x<0||y<0||x>widthPixels||y>heightPixels)
            return 0;
        else
            return Plate[x][y];
    }
    public Bitmap[] getHistoryBitmap(){
        Bitmap[] bitmaps=new Bitmap[Plates.size()];
        return Plates.toArray(bitmaps);
    }
    public void loadHistoryBitmap(){
        Plates.add(getBitmap());
        Log.d(TAG, "loadHistoryBitmap() returned: ");
    }
    public Bitmap getBitmap(){
        Bitmap bitmap=Bitmap.createBitmap(widthPixels,heightPixels,Bitmap.Config.ARGB_8888);
        Log.d("bit",bitmap.getWidth()+","+widthPixels+","+Plate.length);
        for (int ii = 0; ii < bitmap.getWidth(); ii++)
            for (int ii1 = 0; ii1 < bitmap.getHeight(); ii1++) {
                bitmap.setPixel(ii, ii1, get(ii, ii1));
            }
        return bitmap;
    }
    public void setInitCloth(int color_A,int color_B){
        for (int i = 0; i < Plate.length; i++)
            for (int i1 = 0; i1 < Plate[i].length; i1++)
                if(i%2==0)
                    if(i1%2==0)
                        set(i,i1,color_A);
                    else
                        set(i,i1,color_B);
                else
                    if(i1%2==0)
                        set(i,i1,color_B);
                    else
                        set(i,i1,color_A);
    }
    public void setInitBitmap(Bitmap bitmap){
        isWeb=false;
        this.bitmap=bitmap;
        setWidthPixels(bitmap.getWidth());
        setHeightPixels(bitmap.getHeight());
        Log.d("create",bitmap.getWidth()+","+Plate[1].length);
        updateSize();
        for (int i = 0; i < bitmap.getWidth(); i++)
            for (int i1 = 0; i1 < bitmap.getHeight(); i1++) {
            set(i, i1, bitmap.getPixel(i, i1));
        }
    }
    public void updateBitmap(Bitmap bitmap){
        for (int i = 0; i < bitmap.getWidth(); i++)
            for (int i1 = 0; i1 < bitmap.getHeight(); i1++) {
                set(i, i1, bitmap.getPixel(i, i1));
            }
    }
    public void setOnPixelClickListener(final OnPixelClickListener pixelClickListener){
        if(pixelClickListener!=null) {
            setOnTouchListener(new OnTouchListener() {
                @Override
                public boolean onTouch(View view, MotionEvent motionEvent) {
                    PixelPicView instance = PixelPicView.this;
                    motionEvent.recycle();
                    if (motionEvent.getAction() == MotionEvent.ACTION_BUTTON_PRESS) {
                        for (int i = 0; i < Plate.length; i++)
                            for (int i1 = 0; i1 < Plate[i].length; i1++) {
                                if (motionEvent.getX() > i * (getMeasuredWidth()/getWidthPixels())
                                        && motionEvent.getX() < (i + 1) * (getMeasuredWidth()/getWidthPixels())
                                        && motionEvent.getY() > i1 * (getMeasuredHeight()/getHeightPixels())
                                        && motionEvent.getY() < (i1 + 1) * (getMeasuredHeight()/getWidthPixels())) {
                                    pixelClickListener.onClick(instance, i, i1);
                                }
                            }
                    }
                    return true;
                }
            });
        }else setOnTouchListener(null);
    }
    public void setOnPixelTouchListener(@Nullable final OnPixelTouchListener pixelTouchListener) {
        if (pixelTouchListener != null) {
            setOnTouchListener(new OnTouchListener() {
                @Override
                public boolean onTouch(View view, MotionEvent motionEvent) {
                    PixelPicView instance = PixelPicView.this;
                    for (int i = 0; i < Plate.length; i++)
                        for (int i1 = 0; i1 < Plate[i].length; i1++) {
                            if (motionEvent.getX() > i * (getMeasuredWidth()/getWidthPixels())
                                    && motionEvent.getX() < (i + 1) * (getMeasuredWidth()/getWidthPixels())
                                    && motionEvent.getY() > i1 * (getMeasuredHeight()/getHeightPixels())
                                    && motionEvent.getY() < (i1 + 1) * (getMeasuredHeight()/getHeightPixels())) {
                                pixelTouchListener.onTouch(instance, motionEvent, i, i1);
                            }
                        }
                    return true;
                }
            });
        }else setOnTouchListener(null);
    }
    private static float dip2px(float dp){
        return MA_INSTANCE.dip2px(dp);
    }

}

File: app/src/main/java/com/dream/pixeldraw/view/PixelPicView.kt
package com.dream.pixeldraw.view

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.*
import android.graphics.drawable.Drawable
import android.os.Build
import android.util.AttributeSet
import android.util.Log
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.widget.FrameLayout
import android.widget.Toast
import androidx.annotation.RequiresApi
import com.dream.pixeldraw.AppGlobalData.MA_INSTANCE
import com.dream.pixeldraw.AppGlobalData.Plates
import com.dream.pixeldraw.AppGlobalData.Saved_Plates
import com.dream.pixeldraw.AppGlobalData.copied_pic
import java.util.*


open class PixelPicView : View {
    constructor(ctx: Context?) : super(ctx) {}
    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {}
    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}

    private val canvas = Canvas()
    var mWidthPixels = 16
    var mHeightPixels = 16
    private var Plate = arrayOf(intArrayOf(0))
    private var selected = Array(16) { BooleanArray(16) }
    private var isWeb = false
    internal var isSelected: Boolean = false
    internal var background: Drawable? = null
    var mod = 0

    abstract class OnPixelClickListener {
        open fun onClick(view: View?, x: Int, y: Int) {}
    }

    abstract class OnPixelTouchListener {
        open fun onTouch(view: View?, motionEvent: MotionEvent?, x: Int, y: Int){}
    }

    protected fun updateSize() {
        Plate = Array(mWidthPixels) { IntArray(mHeightPixels) }
        selected = Array(mWidthPixels) { BooleanArray(mHeightPixels) }
        for (i in Plate.indices) for (i1 in Plate[i].indices) Plate[i][i1] = Color.TRANSPARENT
        val l = layoutParams as FrameLayout.LayoutParams
        l.width = MA_INSTANCE.dip2px(300f)
        l.height =
            (MA_INSTANCE.dip2px(300f) * (mHeightPixels.toFloat() / mWidthPixels.toFloat())).toInt()
        l.gravity = Gravity.CENTER
        x = (MA_INSTANCE.displayMetrics.widthPixels / 2 - measuredWidth / 2).toFloat()
        y = (MA_INSTANCE.displayMetrics.heightPixels / 2 - measuredHeight / 2).toFloat()
        scaleX = 1f
        scaleY = 1f
        layoutParams = l
        updateCanvas()
    }

    open fun updateCanvas() {
        invalidate()
    }

    override fun invalidate() {
        super.invalidate()
    }

    @SuppressLint("DrawAllocation")
    @RequiresApi(Build.VERSION_CODES.O)
    override fun onDraw(canvas: Canvas) {
        val mpaint2 = Paint()
        mpaint2.isAntiAlias = false
        mpaint2.color = Color.argb(0.5f, 1f, 1f, 1f)
        for (i in Plate.indices) for (i1 in Plate[i].indices) {
            val mpaint = Paint()
            mpaint.isAntiAlias = false
            mpaint.color = Plate[i][i1]
            canvas.drawRect(
                (this.measuredWidth / Plate.size * i).toFloat(),
                (this.measuredHeight / Plate[0].size * i1).toFloat(),
                (this.measuredWidth / Plate.size * (i + 1)).toFloat(),
                (this.measuredHeight / Plate[0].size * (i1 + 1)).toFloat(),
                mpaint
            )
            if (isSelected) {
                val paint = Paint()
                paint.style = Paint.Style.STROKE
                paint.color = Color.argb(0.5f, 1f, 1f, 1f)
                paint.strokeWidth = 8f
                paint.pathEffect = DashPathEffect(floatArrayOf(18f, 8f), 0f)
                if (selected[i][i1]) canvas.drawRect(
                    (this.measuredWidth / Plate.size * i).toFloat(),
                    (this.measuredHeight / Plate[0].size * i1).toFloat(),
                    (this.measuredWidth / Plate.size * (i + 1)).toFloat(),
                    (this.measuredHeight / Plate[0].size * (i1 + 1)).toFloat(),
                    paint
                )
            }
        }
        if (isWeb) {
            for (i in 1 until Plate.size) canvas.drawRect(
                (this.measuredWidth / Plate.size * i).toFloat(),
                0f,
                this.measuredWidth / Plate.size * i + measuredWidth / getWidthPixels() * 0.05f,
                this.measuredHeight.toFloat(),
                mpaint2
            )
            for (i in 1 until Plate[0].size) canvas.drawRect(
                0f,
                this.measuredHeight / Plate[0].size * i - measuredHeight / getHeightPixels() * 0.05f,
                measuredWidth.toFloat(),
                (this.measuredHeight / Plate[0].size * i).toFloat(),
                mpaint2
            )
        }
        val paint = Paint()
        paint.color = Color.argb(0.5f, 1f, 1f, 1f)
        paint.style = Paint.Style.STROKE
        paint.strokeWidth = 4f
        canvas.drawRect(0f, 0f, this.measuredWidth.toFloat(), this.measuredHeight.toFloat(), paint)
        super.onDraw(canvas)
    }

    open fun enableWeb(b: Boolean) {
        if (isWeb != b) {
            isWeb = b
            Log.d("c", "" + b)
            updateCanvas()
        }
    }

    open fun setWidthPixels(count: Int) {
        mWidthPixels = count
        updateSize()
    }

    open fun setHeightPixels(count: Int) {
        mHeightPixels = count
        updateSize()
    }

    open fun getWidthPixels(): Int {
        return mWidthPixels
    }

    open fun getHeightPixels(): Int {
        return mHeightPixels
    }

    open operator fun set(x: Int, y: Int, value: Int) {
        if (x >= 0 && x < mWidthPixels && y >= 0 && y < mHeightPixels) if (Plate[x][y] != value) Plate[x][y] =
            value
        invalidate()
    }

    open operator fun set(x: Float, y: Float, value: Int) {
        if (x >= 0 && x < mWidthPixels && y >= 0 && y < mHeightPixels) if (Plate[x.toInt()][y.toInt()] != value) Plate[x.toInt()][y.toInt()] =
            value
        invalidate()
    }

    open fun drawBitmapAt(x: Int, y: Int, bitmap: Bitmap) {
        for (i in x until x + bitmap.width) for (ii in y until y + bitmap.height) {
            if (i >= 0 && i < getWidthPixels()) if (ii >= 0 && ii < getHeightPixels()) if (bitmap.getPixel(
                    i - x,
                    ii - y
                ) == Color.TRANSPARENT
            ) {
                set(i, ii, bitmap.getPixel(i - x, ii - y))
            }
        }
    }

    open fun selectPixel(x: Int, y: Int) {
        if (x >= 0 && x < mWidthPixels && y >= 0 && y < mHeightPixels) selected[x][y] = true
    }

    open fun renderSelectedPixels() {
        isSelected = true
        updateCanvas()
    }

    open fun cleanSelectedPixels() {
        isSelected = false
        selected = Array(mWidthPixels) { BooleanArray(mHeightPixels) }
        updateCanvas()
    }

    open fun selectRectPixel(x: Int, y: Int, x_end: Int, y_end: Int) {
        for (i in x until x_end) for (ii in y until y_end) selectPixel(i, ii)
    }

    open operator fun get(x: Int, y: Int): Int {
        return if (x < 0 || y < 0 || x > mWidthPixels || y > mHeightPixels) 0 else Plate[x][y]
    }

    open fun loadHistoryBitmap() {
        Plates.add(getBitmap())
        if (MA_INSTANCE.undoButton.visibility == VISIBLE) {
            MA_INSTANCE.undoButton.visibility = INVISIBLE
        }
        Saved_Plates = ArrayList<Bitmap>()
        MA_INSTANCE.returnButton.visibility = VISIBLE
        if (Saved_Plates.size != 0) {
            Saved_Plates = ArrayList<Bitmap>()
        }
    }

    open fun sliceAsBitmap(x: Int, y: Int, end_x: Int, end_y: Int): Bitmap? {
        val bitmap =
            Bitmap.createBitmap(Math.abs(x - end_x), Math.abs(y - end_y), Bitmap.Config.ARGB_8888)
        for (ii in 0 until Math.abs(x - end_x)) for (ii1 in 0 until Math.abs(y - end_y)) {
            bitmap.setPixel(ii, ii1, get(x + ii, y + ii1))
        }
        return bitmap
    }

    open fun getBitmap(): Bitmap {
        val bitmap = Bitmap.createBitmap(mWidthPixels, mHeightPixels, Bitmap.Config.ARGB_8888)
        for (ii in 0 until bitmap.width) for (ii1 in 0 until bitmap.height) {
            bitmap.setPixel(ii, ii1, get(ii, ii1))
        }
        return bitmap
    }

    open fun setInitBitmap(bitmap: Bitmap) {
        isWeb = false
        setWidthPixels(bitmap.width)
        setHeightPixels(bitmap.height)
        Log.d("create", bitmap.width.toString() + "," + Plate[1].size)
        updateSize()
        for (i in 0 until bitmap.width) for (i1 in 0 until bitmap.height) {
            set(i, i1, bitmap.getPixel(i, i1))
        }
    }

    open fun updateBitmap(bitmap: Bitmap) {
        for (i in 0 until bitmap.width) for (i1 in 0 until bitmap.height) {
            set(i, i1, bitmap.getPixel(i, i1))
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    open fun setOnPixelClickListener(pixelClickListener: OnPixelClickListener?) {
        if (pixelClickListener != null) {
            setOnTouchListener { _, motionEvent ->
                val instance = this@PixelPicView
                motionEvent.recycle()
                if (motionEvent.action == MotionEvent.ACTION_BUTTON_PRESS) {
                    for (i in Plate.indices) for (i1 in 0 until Plate[i].size) {
                        if (motionEvent.x > i * (measuredWidth / getWidthPixels()) && motionEvent.x < (i + 1) * (measuredWidth / getWidthPixels()) && motionEvent.y > i1 * (measuredHeight / getHeightPixels()) && motionEvent.y < (i1 + 1) * (measuredHeight / getWidthPixels())
                        ) {
                            pixelClickListener.onClick(instance, i, i1)
                        }
                    }
                }
                true
            }
        } else setOnTouchListener(null)
    }

    @SuppressLint("ClickableViewAccessibility")
    open fun setOnPixelTouchListener(pixelTouchListener: OnPixelTouchListener?) {
        if (pixelTouchListener != null) {
            setOnTouchListener(object : OnTouchListener {
                var down: Long = 0
                override fun onTouch(view: View, motionEvent: MotionEvent): Boolean {
                    val instance: PixelPicView = this@PixelPicView
                    for (i in Plate.indices) for (i1 in Plate[i].indices) {
                        if (motionEvent.x > i * (measuredWidth / getWidthPixels()) && motionEvent.x < (i + 1) * (measuredWidth / getWidthPixels()) && motionEvent.y > i1 * (measuredHeight / getHeightPixels()) && motionEvent.y < (i1 + 1) * (measuredHeight / getHeightPixels())) {
                            pixelTouchListener.onTouch(instance, motionEvent, i, i1)
                            if (motionEvent.action == MotionEvent.ACTION_DOWN) down = Date().time
                            if (motionEvent.action == MotionEvent.ACTION_UP) if (Date().time - down >= 400 && copied_pic != null) {
                                Toast.makeText(context, "已粘贴", Toast.LENGTH_SHORT).show()
                                drawBitmapAt(i, i1, copied_pic!!)
                            }
                        }
                    }
                    return true
                }
            })
        } else setOnTouchListener { _, _ ->
            setOnTouchListener(object : OnTouchListener {
                var down: Long = 0
                override fun onTouch(
                    view: View,
                    motionEvent: MotionEvent
                ): Boolean {
                    for (i in Plate.indices) for (i1 in Plate[i].indices) {
                        if (motionEvent.x > i * (measuredWidth / getWidthPixels()) && motionEvent.x < (i + 1) * (measuredWidth / getWidthPixels()) && motionEvent.y > i1 * (measuredHeight / getHeightPixels()) && motionEvent.y < (i1 + 1) * (measuredHeight / getHeightPixels())
                        ) {
                            if (motionEvent.action == MotionEvent.ACTION_DOWN) down =
                                Date().time
                            if (motionEvent.action == MotionEvent.ACTION_UP) if (Date().time - down >= 400 && copied_pic != null) {
                                Toast.makeText(context, "已粘贴", Toast.LENGTH_SHORT).show()
                                drawBitmapAt(i, i1, copied_pic!!)
                            }
                        }
                    }
                    return true
                }
            })
            true
        }
    }
}